build-tools/packages/build-tools/src/typeValidator/typeData.ts

```diff
@@ -99,8 +99,10 @@ export function toTypeString(prefix: string, typeData: TypeData) {
 		) {
 			// it's really hard to build the right type for a generic,
 			// so for now we'll just pass any, as it will always work
+			// even though it may defeat the utility of a type or related test.
 			typeParams = `<${node
 				.getTypeParameters()
+				.filter((tp) => tp.getDefault() === undefined)
```

Comment:
does this only work because required as required params can't follow optional params, e.g. `type foo<A=string,B,C>`

how does this work for a type like `type foo<A extends string,B,C>`

Reply:
That is correct. Optional parameters (those with defaults) must follow any required (no default) parameters.

This doesn't change anything for constrained parameters (has `extends`). Unfortunately when `any` is forced on a type parameter that has a constraint the result is still `any`. I agree matching constraint has complications.